bftengine/src/bcstatetransfer/Messages.hpp

```diff
@@ -40,56 +40,55 @@ class MsgType {
 };
 
 struct BCStateTranBaseMsg {
+  BCStateTranBaseMsg(uint16_t type) : type(type) {}
   uint16_t type;
+  // This struct and its derived structs are used to de/serialize message buffers sent over communication channels.
+  // Since virtual methods modify the memory layout of a struct, there cannot be any for both this base struct and its
+  // inheritors.
 };
 
-struct AskForCheckpointSummariesMsg : public BCStateTranBaseMsg {
-  AskForCheckpointSummariesMsg() {
-    memset(this, 0, sizeof(AskForCheckpointSummariesMsg));
-    type = MsgType::AskForCheckpointSummaries;
+// MsgAllocator is useful for structs with a flexible array member. T should have a method setDataSize.
+template <typename T>
+class MsgAllocator {
+ public:
+  MsgAllocator(size_t dataSize) : memory_(new concord::Byte[calcBufferSize(dataSize)]()) {
+    (*this).operator->()->setDataSize(dataSize);
   }
 
-  uint64_t msgSeqNum;
-  uint64_t minRelevantCheckpointNum;
+  T* operator->() const { return reinterpret_cast<T*>(memory_.get()); }
+
+  concord::Byte* getBytes() const { return memory_.get(); }
+
+  static size_t calcBufferSize(size_t dataSize) { return sizeof(T) - 1 + dataSize; }
+
+ private:
+  std::unique_ptr<concord::Byte[]> memory_;
```

Comment:
Calling an array of bytes "memory" isn't standard..I suggest to switch to buff_/buffer_ which is more broadly used for this purpose.
On a 2nd though data_ is even better.

Reply:
The previously used function is called malloc - **memory** is in the name of the standard function for allocations. This class handles the allocation of messages.

Reply:
I believe that data_ will be a confusing name as we use this name as member data[1] and dataSize which refers to that member.